Bug 1290331 - Removed incorrect and confusing .NotStarted sync warning messaging

sleroux opened the pull request:
When the sync operations complete, we are erroneously interpreting .NotStarted sync results caused by users disabling a particular sync option (tabs, history, etc) as 'Sync is unavailable'. Related, in the case the user is offline, we show the message 'Sync is offline'. The first case is not correct and we should reporting a good sync. In the second, the messaging is unclear. Are the sync servers offline? I think the messaging can be improved here. In the meantime, I propose we remove both checks until we improve the messaging/UI around reporting sync errors.

farhanpatel commented on Providers/Profile.swift:
```diff
@@ -958,21 +958,16 @@ public class BrowserProfile: Profile {
-            case .NotStarted(let reason):
-                let message: String
-                switch reason {
-                case .Offline:
```
should there be a tab for this switch statement?

farhanpatel commented on Providers/Profile.swift:
```diff
@@ -958,21 +958,16 @@ public class BrowserProfile: Profile {
-                    message = Strings.FirefoxSyncNotStartedTitle
-                }
-                return SyncDisplayState.Stale(message: message)
+                return .Good
```
If the default case is to return .Good. Should we handle the happy case here explicitly? If not we could remove the .Complete case.

sleroux replied:
Ya I can probably remove the explicit case and just use the default for now.

farhanpatel commented on Providers/Profile.swift:
```diff
@@ -958,21 +958,16 @@ public class BrowserProfile: Profile {
                     return SyncDisplayState.Stale(message: String(format:Strings.FirefoxSyncPartialTitle, Strings.localizedStringForSyncComponent(identifier) ?? ""))
                 }
-                return SyncDisplayState.Stale(message: Strings.FirefoxSyncNotStartedTitle)
+                return .Stale(message: Strings.FirefoxSyncNotStartedTitle)
```
I'm assuming .Partial is a partial sync. In that case it looks like we are still returning the same message NotStartedTitle

sleroux replied:
I'm actually not sure if .Partial is something we want to show a warning about either. The messaging should probably not be NotStarted though since the operation did start - it just partially completed.

farhanpatel replied:
Yea good point. A partial sync for a user who only enabled bookmarks sync is still a complete sync in their point of view. Best to just show complete in that case

farhanpatel commented on Providers/Profile.swift:
```diff
@@ -958,21 +958,16 @@ public class BrowserProfile: Profile {
             switch syncStatus {
             case .Completed:
-                return SyncDisplayState.Good
-            case .NotStarted(let reason):
```
By removing the .NotStarted case wont the browser just default to Good even when the sync has failed. You mentioned that the SyncOfflineTitle was wrong. But in the case of the sync not starting for various reasons the NotStartedTitle would still be helpful instead of making them think sync succeeded.

ncalexan commented:
I feel like this is addressing the problem at the wrong level.  What should happen is that each sync engine should surface it's state explicitly.  Then it's up to the thing that reduces all of those states into one global state to display properly.  That is, the sync engines should surface the state `.DisabledRemotely` (which may be new), and then the reduction should handle that like `.Good`.

I haven't reread this code entirely, but it looks like there are two places the Sync state is displayed in the settings, and both need to handle `.DisabledRemotely` for engines appropriately.

@sleroux is that cogent?

sleroux commented:
Thanks for taking a look @ncalexan!

> I feel like this is addressing the problem at the wrong level. What should happen is that each sync engine should surface it's state explicitly.

A bit further up the stack we receive an EngineResults object which contains a result for each sync engine [1]. The reducing of the states into a single state happens one stack frame up in `displayStateForEngineResults`. 

> That is, the sync engines should surface the state .DisabledRemotely

We have something similar we might be able to use. The `SyncStatus.NotStarted` enum has an associated reason for why it hasn't started. One reason being `EngineRemotelyNotEnabled` [2]. We can use this to better deduce why we never started the sync operation for a particular collection.

> @sleroux is that cogent?

Pretty much. We can determine what to show more intelligently by digging into more of the details of each sync and reducing into a state that makes sense for the user instead of this more brute approach of saying everything is good. 

Other than `.EngineRemotelyNotEnabled`, are there any other NotStarted or Partial states that should not result in a warning? The original approach of saying anything not `.NotStarted` or `.Partial` results in an error isn't working. I'd imagine that Android has similar logic to determine what to display when there are issues with syncing. I wonder if it's possible to translate that into our states. 

[1] https://github.com/mozilla/firefox-ios/blob/master/Providers/Profile.swift#L684
[2] https://github.com/mozilla/firefox-ios/blob/master/Sync/Synchronizers/Synchronizer.swift#L103

rnewman commented:
Concur that the right thing to do here is not to remove granularity, it's to correctly represent and handle state.

rnewman commented:
I would go so far as to say you should `switch` over all of the result enums and decide which ones you do want to show messaging for. Some such states will be a problem indicator only if they are prolonged.

sleroux commented:
@farhanpatel I've updated the PR if you want to take another look or if @rnewman / @ncalexan want to feel free.

> I would go so far as to say you should switch over all of the result enums and decide which ones you do want to show messaging for. Some such states will be a problem indicator only if they are prolonged.

I was thinking along the same lines. The latest commit I pushed up switches on the various reasons at a more fine grained level to determine what the display status should be. Since the UI only supports showing a single state at this time, I reduce all the errors and pass back the last, worst error. For example, if all engines were successful except for one that was stale and another was offline, the aggregate display state would be the offline one.

I've also extracted out the sync state resolving code into its own struct for now for easier unit testing. I'm not 100% clear on what each reason should return (warning vs error) so feel free to comment inline if I've missed some.

farhanpatel commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+
+public func ==(a: SyncDisplayState, b: SyncDisplayState) -> Bool {
+    switch (a, b) {
+    case (.InProgress,   .InProgress): return true
```
nit: This should be tabbed

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+
+private let log = Logger.syncLogger
+
+public struct SyncStatusResolver {
```
Block comment explaining what this does: translate between fine-grained results of sync operations and coarse-grained display-oriented states for error display.

Future: perhaps integrating statuses over time ("we haven't been able to sync your bookmarks for seven days").

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+            case .NotStarted(let reason):
+                switch reason {
+
+                // Potentially problematic reasons for not starting. Should surface to the user.
```
I feel like some of these depend on the application state. If I tap "Sync Now" and anything bad happens, tell me. An automatic sync shouldn't complain because I'm offline!

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+                case .Offline:
+                    return .Bad(message: Strings.FirefoxSyncOfflineTitle)
+                case .NoAccount:
+                    // Should this ever happen? Why would we sync with no account?
```
If we're syncing while the user deletes their account, or presumably if we have a bug that loses some account state.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+                default:
+                    return .Good
+                }
+            default:
```
I would rather see a verbose listing of every case. One of the strengths of Swift's enums is that the compiler will warn you if you missed a case… but it can't do that if you fall back to `default`. If we add more statuses, we'll want that help.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        // TODO: Instead of finding the worst offender in a list of statuses, we should better surface
+        // what might have happend with a particular engine when syncing.
```
s/happend/happened

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+                return displayState
+
+            case .Stale(_):
+                // If the state we're carrying is worst than the stale one, keep passing
```
s/worst/worse

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+    case InProgress
+    case Good
+    case Bad(message: String?)
+    case Stale(message: String)
```
What does this really mean? Would benefit from a comment.

sleroux replied:
I've renamed this to Warning since that what we display to the user. I don't think the meaning of Stale makes sense here.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+    public func resolveResults() -> SyncDisplayState {
+        guard let results = engineResults.successValue else {
+            switch engineResults.failureValue {
+            case _ as BookmarksMergeError, _ as BookmarksDatabaseError:
```
`BookmarksDatabaseError` is probably misnamed at this point. My bad.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,110 @@
+            case _ as BookmarksMergeError, _ as BookmarksDatabaseError:
+                return SyncDisplayState.Stale(message: String(format: Strings.FirefoxSyncPartialTitle, Strings.localizedStringForSyncComponent("bookmarks") ?? ""))
+            default:
+                return SyncDisplayState.Bad(message: nil)
```
What other kinds of errors can we get here? How about things like interrupted network connections (host not found, etc.), or TLS errors from captive portals?

I'm tempted to default to not messaging here, and messaging error states for specific actionable issues.

rnewman commented:
This looks good to me, in that it fixes the issue, but you might consider a follow-up to address error handling more carefully.

rnewman commented:
Thumbs up for writing for testability, btw. 👍

sleroux commented:
Updated PR with minor nits and fixes. Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1291718 for follow up to handle errors better with contextual info/time.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,136 @@
+                case .EngineFormatTooNew(_):
+                    return .Good
+                case .StorageFormatOutdated(_):
+                    return .Good
```
In this case you need to update your version of Firefox for iOS.

rnewman commented on Providers/SyncStatusResolver.swift:
```diff
@@ -0,0 +1,136 @@
+                case .EngineRemotelyNotEnabled(_):
+                    return .Good
+                case .EngineFormatOutdated(_):
+                    return .Good
```
In this case you need to update your version of Firefox for iOS.

rnewman commented:
LGTM.
